src/social.ts

```diff
@@ -0,0 +1,241 @@
+            // meta: {},
+            lastUpdated: 0, // timestamp is updated when this contact's social graph is updated
+          }
+        }
+      }
+    }
+
+    // we just updated all the contacts for pubkey; update its lastUpdated timestamp
+    if (Object.prototype.hasOwnProperty.call(this.socialGraph, pubkey)) {
+      this.socialGraph[pubkey].lastUpdated = Date.now()
+    }
+
+    // this.analyze() // console log stuff about the graph as it is built.
+
+    this.iterateGraph()
+
+  }
+  /**
+   * Iterate over the social graph and update each contact's social graph if it is stale
+   * Iteration should not move to the next degree until all contacts in the current degree have been updated - breadth first instead of depth first for good UX.
+   */
+  iterateGraph() {
+    const now = Date.now()
+    const graphKeys = Object.keys(this.socialGraph)
+    if (this.iteration >= graphKeys.length) {
+      // we've completed the graph. start over
+      this.iteration = 1
+    }
+    // get contact
+    const contact = graphKeys[this.iteration]
+    // if the contact's social graph is stale and < GRAPH_DEPTH degree, update it
+    if (
+      now - this.socialGraph[contact].lastUpdated > STALE_GRAPH && 
+      this.socialGraph[contact].degree + 1 <= GRAPH_DEPTH
+    ) {
+      this.extendGraph(contact, this.socialGraph[contact].degree + 1)
+      this.iteration++
+    } else {
+      // keep iterating
+      // once per second to avoid resource hogging 
+      this.iteration++
+      setTimeout(this.iterateGraph,1000)
+    }
+  }
+  /**
+   * show when each contact was last updated
+   * currently unused outside of the console
+   * @returns an array of pubkeys sorted by lastUpdated
+   */
+  lastUpdate(){
+    return Object.keys(this.socialGraph).sort( (a,b) => this.socialGraph[b].lastUpdated - this.socialGraph[a].lastUpdated ).map( pk => `${(((+new Date()) - this.socialGraph[pk].lastUpdated)/1000/60).toFixed(1) + 'm ago'} - 0x${pk.substring(0,6)} ` )
+  }
+  /**
+   * analyze the social graph as it is built
+   * Debug console only.
+   */
+  // analyze(){
+  //   var degrees = {};
+  //   Object.keys(this.socialGraph).map(s => degrees[this.socialGraph[s].degree] = degrees[this.socialGraph[s].degree] ? degrees[this.socialGraph[s].degree] + 1 : 1)
+
+  //   console.log('count of each degree', degrees)
+
+  //   // sort top 10 fwf contacts
+  //   // return their pubkey and fwf value
+  //   const top10 = Object.keys(this.socialGraph)
+  //     .map(s => this.socialGraph[s])
+  //     .sort((a, b) => b.fwf - a.fwf)
+  //     .slice(0, 10)
+  //     .map(s => ({ pubkey: s.pubkey, fwf: s.fwf }))
+  //   console.log('top 10 fwf contacts', top10)
+
+  // }
+  // given a pubkey, return the pubkeys between it and the root or an empty array if none
+  // the pubkeys will be in the order of closest to your pubkey to closest to the target pubkey
+  traverse(pubkey: string) {
```

Comment:
```
// can you do these, and add to test?

distance(pubkey) {
   if (pubkey == root) return 0;
   if (pubkey not in this graph) return high_number;
   if (pubkey in this graph) traverse(pubkey).length) + 1
}

weight(pubkey) {
   if distance == 0l return high_numberl
   return 1 / distance * distance;
}
```

Reply:
I was unable to push to the social-graph branch anymore... permissions changed? Here is my new PR from a fork: https://github.com/ArcadeLabsInc/arclib/pull/36 @earonesty